File: clients/python/beneath/schema.py
from collections.abc import Mapping
import json
import io
from typing import Tuple

from fastavro import parse_schema
from fastavro import schemaless_reader
from fastavro import schemaless_writer

from beneath.proto import gateway_pb2
from beneath.utils import (
    ms_to_datetime,
    ms_to_pd_timestamp,
    timestamp_to_ms,
)


class Schema:
    """ Represents a stream's parsed Avro schema """

    def __init__(self, avro: str):
        self.parsed_avro = parse_schema(json.loads(avro))
        """ The parsed avro schema """

    def record_to_pb(self, record: Mapping) -> Tuple[gateway_pb2.Record, int]:
        if not isinstance(record, Mapping):
            raise TypeError("write error: record must be a mapping, got {}".format(record))
        avro = self._encode_avro(record)
        timestamp = self._extract_record_timestamp(record)
        pb = gateway_pb2.Record(avro_data=avro, timestamp=timestamp)
        return (pb, pb.ByteSize())

    def pb_to_record(self, pb: gateway_pb2.Record, to_dataframe: bool) -> Mapping:
        record = self._decode_avro(pb.avro_data)
        record["@meta.timestamp"] = (
            ms_to_pd_timestamp(pb.timestamp) if to_dataframe else ms_to_datetime(pb.timestamp)
        )
        return record

    def _encode_avro(self, record: Mapping):
        writer = io.BytesIO()
        schemaless_writer(writer, self.parsed_avro, record)
        result = writer.getvalue()
        writer.close()
        return result

    def _decode_avro(self, data):
        reader = io.BytesIO(data)
        record = schemaless_reader(reader, self.parsed_avro)
        reader.close()
        return record

    @staticmethod
    def _extract_record_timestamp(record: Mapping) -> int:
        if ("@meta" in record) and ("timestamp" in record["@meta"]):
            return timestamp_to_ms(record["@meta"]["timestamp"])
        return 0  # 0 tells the server to set timestamp to its current time

File: clients/java/src/main/java/dev/beneath/client/Schema.java
package dev.beneath.client;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Map;
import java.util.Map.Entry;

import com.google.protobuf.ByteString;

import org.apache.avro.generic.GenericDatumReader;
import org.apache.avro.generic.GenericDatumWriter;
import org.apache.avro.generic.GenericRecord;
import org.apache.avro.io.BinaryEncoder;
import org.apache.avro.io.DatumReader;
import org.apache.avro.io.DatumWriter;
import org.apache.avro.io.Decoder;
import org.apache.avro.io.DecoderFactory;
import org.apache.avro.io.EncoderFactory;

/**
 * Represents a table's parsed Avro schema
 */
public class Schema {
  public String avro;
  public org.apache.avro.Schema parsedAvro;

  public Schema(String avro) {
    this.avro = avro;
    this.parsedAvro = new org.apache.avro.Schema.Parser().parse(avro);
  }

  // TODO: what's the best way to return a tuple?
  public Entry<Record, Integer> recordToPb(GenericRecord record) {
    byte[] avro = this.encodeAvro(record);
    Integer timestamp = extractRecordTimestamp(record);
    Record pb = Record.newBuilder().setAvroData(ByteString.copyFrom(avro)).setTimestamp(timestamp).build();
    return Map.entry(pb, pb.getSerializedSize());
  }

  public GenericRecord pbToRecord(Record pb) {
    GenericRecord record = this.decodeAvro(pb.getAvroData().toByteArray());
    // record.put("@meta.timestamp", "not yet implemented");
    return record;
  }

  private byte[] encodeAvro(GenericRecord record) {
    ByteArrayOutputStream out = new ByteArrayOutputStream();
    BinaryEncoder encoder = EncoderFactory.get().binaryEncoder(out, null);
    DatumWriter<GenericRecord> writer = new GenericDatumWriter<GenericRecord>(this.parsedAvro);
    try {
      writer.write(record, encoder);
      encoder.flush();
      out.close();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    byte[] serializedBytes = out.toByteArray();
    return serializedBytes;
  }

  private GenericRecord decodeAvro(byte[] data) {
    DatumReader<GenericRecord> reader = new GenericDatumReader<GenericRecord>(this.parsedAvro);
    Decoder decoder = DecoderFactory.get().binaryDecoder(data, null);
    GenericRecord record;
    try {
      record = reader.read(null, decoder);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return record;
  }

  private static Integer extractRecordTimestamp(GenericRecord record) {
    // TODO: actually check the record for a timestamp
    return 0; // 0 tells the server to set timestamp to its current time
  }
}
